File: Matlab/calculus/find_root/func/func_func_newton_raphson.m
% Start

%{
func = @(x)(x.^6 - x - 1);
Dfunc = @(x)(6.*x.^5 - 1);
ax = 1.5;
err = 0.0001;
Niter = 100;
%}

% Function
function [NsolF, EsolF, NsolE] = func_func_newton_raphson(func, Dfunc, ax, err, Niter)

% Input
if nargin < 4
    err = 0.00001;
end
if nargin < 5
    Niter = 100;
end

% Algorithm
for iter = 1:Niter
    af = func(ax);
    adf = Dfunc(ax);
    NsolF = ax - af/adf;
    Nf = func(NsolF);
    if abs(Nf) < err
        break;
    end
    ax = NsolF;
end

% Compare
EsolF = fzero(func, ax);
NsolE = abs(EsolF - NsolF);

% **************************************************^**************************************************
% *****************************# Copyright by Ali Forouzandeh Hafshejani #*****************************
% ******************************# Scientific Computing Specialist 20@21 #******************************
% **************************************************^**************************************************

File: C++/calculus/find_root/func/func_func_newton_raphson.cpp
// Start
#include <complex>
using namespace std;

/*
double func(double x)
{
    return (
        cos(2*x) + sin(2*x) + x - 1
    );
}

double Dfunc(double x)
{
    return (
        2*cos(2*x) - 2*sin(2*x) + 1
    );
}

double ax = 1.5;
double err = 0.00001;
int Niter = 100;
*/

// Function
double func_func_newton_raphson(double (*func)(double), double (*Dfunc)(double), double ax, double err = 0.00001, int Niter = 100)
{
    int iter;
    double af, aDf, NsolF, NvalF;

    // Algorithm
    for (iter = 0; iter < Niter; ++iter)
    {
        af = func(ax);
        aDf = Dfunc(ax);
        NsolF = ax - af/aDf;
        NvalF = func(NsolF);

        if (abs(NvalF) < err)
        {
            break;
        }
        ax = NsolF;
    }

    return NsolF;
}

// **************************************************^**************************************************
// *****************************# Copyright by Ali Forouzandeh Hafshejani #*****************************
// ******************************# Scientific Computing Specialist 20@21 #******************************
// **************************************************^**************************************************